src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,415 @@
+end
+
+lemma not_inc_set_orientation {i : V} {e : G.edge_set}
+  (H_head : i ≠ o.head e) (H_tail : i ≠ o.tail e) : e.val ∉ G.incidence_set i :=
+begin
+  intro h,
+  rw G.incidence_set_orientation at h,
+  tauto,
+end
+
+end orientations
+
+section oriented_incidence
+
+variables [fintype V] [decidable_eq V] [decidable_rel G.adj] (R) [ring R] {o : orientation G}
+
+-- ## Oriented Incidence Matrix N(o)
+
+/-- An `oriented incidence matrix` N(o) is defined with respect to the orientation of the edges
+and is defined to be `1` for entries (`i`,`e`) where `i` is the head of `e`, `-1` where `i`
+is the tail of `e`, and `0` otherwise. -/
+def oriented_inc_matrix (o : orientation G) : matrix V G.edge_set R :=
+λ i e, if i = o.head e then (1 : R) else (if i = o.tail e then -1 else 0)
+
+@[simp]
+lemma oriented_inc_matrix_apply {i : V} {e : G.edge_set} :
+  G.oriented_inc_matrix R o i e = ite (i = o.head e) (1 : R) (ite (i = o.tail e) (-1) 0) := rfl
+
+lemma oriented_inc_matrix_head {i : V} {e : G.edge_set} (H_head : i = o.head e) :
+  G.oriented_inc_matrix R o i e = 1 :=
+by simp only [H_head, if_true, eq_self_iff_true, oriented_inc_matrix_apply]
+
+lemma oriented_inc_matrix_tail {i : V} {e : G.edge_set} (H_tail : i = o.tail e) :
+  G.oriented_inc_matrix R o i e = -1 :=
+by simp only [H_tail, oriented_inc_matrix, (G.head_neq_tail).symm,
+              if_false, if_true, eq_self_iff_true]
+
+lemma oriented_inc_matrix_zero {i : V} {e : G.edge_set} [char_zero R] :
+  G.oriented_inc_matrix R o i e = 0 ↔ i ≠ o.head e ∧ i ≠ o.tail e :=
+begin
+  by_cases H₁ : i = o.head e,
+  { simp only [oriented_inc_matrix, H₁, if_true, eq_self_iff_true, not_true,
+               ne.def, one_ne_zero, false_and] },
+  { by_cases H₂ : i = o.tail e,
+    { simp only [H₂, oriented_inc_matrix_tail, eq_self_iff_true, not_true,
+                 ne.def, neg_eq_zero, one_ne_zero, and_false] },
+    { simp only [H₁, H₂, eq_self_iff_true, if_false, ne.def,
+                 not_false_iff, and_self, oriented_inc_matrix_apply] } }
+end
+
+lemma oriented_inc_matrix_non_zero {i : V} {e : G.edge_set} [char_zero R] :
+  ¬ G.oriented_inc_matrix R o i e = 0 ↔ i = o.head e ∨ i = o.tail e :=
+begin
+  by_cases H₁ : i = o.head e,
+  { simp only [H₁, if_true, true_or, eq_self_iff_true, ne.def,
+               not_false_iff, one_ne_zero, oriented_inc_matrix_apply] },
+  { by_cases H₂ : i = o.tail e,
+    { simp only [H₂, oriented_inc_matrix_tail, eq_self_iff_true, ne.def, or_true,
+                 not_false_iff, neg_eq_zero, one_ne_zero] },
+    { simp only [H₁, H₂, eq_self_iff_true, not_true, if_false,
+                 ne.def, oriented_inc_matrix_apply, or_self] } }
+end
+
+-- 5. (N(o) i e) ^ 2 = M i e, for any orientation o, vertex i and edge e.
+theorem oriented_inc_matrix_elem_squared {i : V} {e : G.edge_set} [char_zero R] :
+  G.oriented_inc_matrix R o i e * G.oriented_inc_matrix R o i e = G.inc_matrix R i e :=
+begin
+  by_cases H_head : i = o.head e,
```

Comment:
it may be worth having a lemma like `i = o.head e ∨ i = o.tail e ∨ <otherwise>`, so `by_cases` can be replaced with a nice `obtain`

Reply:
I am not sure what the result with the `obtain` syntax would look like, since I haven't worked with it yet. Can you please give an example here? :smile:

Reply:
so for example, I'm gonna use [`lt_trichotomy`](https://leanprover-community.github.io/mathlib_docs/init/algebra/order.html#lt_trichotomy), you end up with something like this:

```lean
import data.real.basic

example (x y : ℝ) : false :=
begin
  obtain le | eq | gt := lt_trichotomy x y,
end
```

you end up with three goals depending to each case, which is lovely.

Reply:
yours would probably look like (a ∨ b ∨ ¬a ∧ ¬b), which ends up looking like this:

```lean
import tactic

lemma asda (a b : Prop) : a ∨ b ∨ ¬a ∧ ¬b := by tauto

example (x y : Prop) : false :=
begin
  obtain a | b | ⟨na, nb⟩ := asda x y,
end
```

Reply:
And would it be stupid to have instead of the `<otherwise>` `i ≠ o.head e ∧ i ≠ o.tail e` or should I come up with something else?

Reply:
yeah that's what I imagined it being!